webrender/src/device/gfx/descriptor.rs

```diff
@@ -2,409 +2,519 @@
+
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Debug, Default)]
+pub(super) struct PerPassBindings(pub [TextureId; PER_PASS_TEXTURE_COUNT]);
+
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Debug, Default)]
+pub(super) struct PerGroupBindings(pub [TextureId; PER_GROUP_TEXTURE_COUNT]);
+
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Debug)]
+pub(super) struct SamplerBindings(pub [TextureFilter; MUTABLE_SAMPLER_COUNT]);
+
+impl Default for SamplerBindings {
+    fn default() -> Self {
+        SamplerBindings([TextureFilter::Linear; MUTABLE_SAMPLER_COUNT])
+    }
+}
+
+pub(super) struct DescriptorGroupData<B: hal::Backend> {
+    pub(super) descriptor_set_layouts: FastHashMap<DescriptorGroup, ArrayVec<[B::DescriptorSetLayout; DESCRIPTOR_SET_PER_LAYOUT]>>,
+    pub(super) descriptor_set_ranges: FastHashMap<DescriptorGroup, ArrayVec<[DescriptorRanges; DESCRIPTOR_SET_PER_LAYOUT]>>,
+    pub(super) pipeline_layouts: FastHashMap<DescriptorGroup, B::PipelineLayout>,
+}
+
+impl<B: hal::Backend> DescriptorGroupData<B> {
+    pub(super) fn descriptor_layout(&self, group: &DescriptorGroup, group_idx: usize) -> &B::DescriptorSetLayout {
+        &self.descriptor_set_layouts[group][group_idx]
     }
 
-    pub(super) fn descriptor_set_at_idx(&self, index: usize) -> &B::DescriptorSet {
-        assert!(index <= self.current_descriptor_set_idx);
-        &self.descriptor_set[index]
+    pub(super) fn ranges(&self, group: &DescriptorGroup, group_idx: usize) -> DescriptorRanges {
+        self.descriptor_set_ranges[group][group_idx]
     }
 
-    pub(super) fn descriptor_set_layout(&self) -> &B::DescriptorSetLayout {
-        &self.descriptor_set_layout
+    pub(super) fn pipeline_layout(&self, group: &DescriptorGroup) -> &B::PipelineLayout {
+        &self.pipeline_layouts[group]
     }
 
-    pub(super) fn next(&mut self) -> bool {
-        if !self.free_sets.is_empty() {
-            return true;
-        }
-        self.current_descriptor_set_idx += 1;
-        if self.current_descriptor_set_idx >= self.max_descriptor_set_size {
-            return false;
+    pub(super) unsafe fn deinit(self, device: &B::Device) {
+        for (_, layouts) in self.descriptor_set_layouts {
+            for layout in layouts {
+                device.destroy_descriptor_set_layout(layout);
+            }
         }
-        if self.current_descriptor_set_idx == self.descriptor_set.len() {
-            self.allocate();
+        for (_, layout) in self.pipeline_layouts {
+            device.destroy_pipeline_layout(layout);
         }
-        true
     }
+}
 
-    fn allocate(&mut self) {
-        let desc_set = unsafe {
-            self.descriptor_pool
-                .allocate_set(&self.descriptor_set_layout)
-        }
-        .expect(&format!(
-            "Failed to allocate set with layout: {:?}!",
-            self.descriptor_set_layout,
-        ));
-        self.descriptor_set.push(desc_set);
+pub(super) trait FreeSets<B: hal::Backend> {
```

Comment:
for things like this, it would be good to have a piece of documentation explaining what the trait is for, then on each method explaining what the method is for, specifically